### ISSUE TITLE
PT implementation of Transformer has very bad translation results

### ISSUE THREAD

**Issue by astonzhang** [d2l-ai/d2l-en#1484]
http://preview.d2l.ai.s3-website-us-west-2.amazonaws.com/d2l-en/master/chapter_attention-mechanisms/transformer.html

PyTorch:

go . => <unk> !, bleu 0.000
i lost . => je l'ai vu ., bleu 0.000
so long . => <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk>, bleu 0.000
i'm home . => <unk> à la maison ., bleu 0.000
he's calm . => sois <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk>, bleu 0.000

MXNet:

go . => va !, bleu 1.000
i lost . => j'ai perdu ., bleu 1.000
so long . => <unk> !, bleu 0.000
i'm home . => je suis chez <unk> chez vous en <unk> <unk> <unk>, bleu 0.376
he's calm . => il <unk> calme ., bleu 0.658

Though the translation looks good in other sections:
http://preview.d2l.ai.s3-website-us-west-2.amazonaws.com/d2l-en/master/chapter_recurrent-modern/seq2seq.html
http://preview.d2l.ai.s3-website-us-west-2.amazonaws.com/d2l-en/master/chapter_attention-mechanisms/seq2seq-attention.html

**Comment by AnirudhDagar** [comment]
Hmm, I can't see the root cause here.
If you check the stable version that seems to be doing very well. http://d2l.ai/chapter_attention-mechanisms/transformer.html#training

Could it be just this run or the change of hyperparams?  I'll try to investigate more.

**Comment by astonzhang** [comment]
In the stable version, both mx & pt perform poorly because there are issues in the initial implementations. Now they're fixed, but the pt version still didn't perform as well as the mx version:

http://preview.d2l.ai.s3-website-us-west-2.amazonaws.com/d2l-en/master/chapter_attention-mechanisms/transformer.html

MXNet:
go . => va !, bleu 1.000
i lost . => j'ai perdu ., bleu 1.000
i'm home . => je suis chez moi ., bleu 1.000
he's calm . => il est malade ., bleu 0.658

PyTorch:
go . => c'est le <unk> !, bleu 0.000
i lost . => je suis paresseuse ., bleu 0.000
i'm home . => je suis chez moi ., bleu 1.000
he's calm . => je suis calme ., bleu 0.537

**Comment by astonzhang** [comment]
One hint is that 
mx: loss 0.030
pt: loss 0.016

These are quite different

**Comment by AnirudhDagar** [comment]
I checked again and PyTorch is actually performing better than MX in this case if you look closely bleu score for the 1st example is perfect in PT unlike MX.

MX
```
go . => va le chercher !, bleu 0.000
i lost . => j'ai perdu ., bleu 1.000
i'm home . => je suis chez moi ., bleu 1.000
he's calm . => il est bon ., bleu 0.658
```

PT
```
go . => va !, bleu 1.000
i lost . => j'ai perdu ., bleu 1.000
i'm home . => je suis <unk> ., bleu 0.512
he's calm . => il est mouillé ., bleu 0.658
```

The loss difference is not a lot if you are hinting towards overfitting in PT.

**Comment by astonzhang** [comment]
Thanks. Could you run experiments for multiple runs to see if mx and pt results are comparable?

**Comment by AnirudhDagar** [comment]
Run 1: `loss 0.014, 5796.9 tokens/sec on cpu`
```
go . => entrez !, bleu 0.000
i lost . => j'ai perdu ., bleu 1.000
i'm home . => j'en suis chez moi ., bleu 0.832
he's calm . => il est calme ., bleu 1.000
```

Run 2: `loss 0.014, 5826.1 tokens/sec on cpu`
```
go . => <unk> !, bleu 0.000
i lost . => je suis perdu ., bleu 0.537
i'm home . => je suis chez moi ., bleu 1.000
he's calm . => je suis calme ., bleu 0.537
```

Run 3: `loss 0.016, 5568.3 tokens/sec on cpu`
```
go . => <unk> !, bleu 0.000
i lost . => je sais perdu ., bleu 0.537
i'm home . => je suis chez moi ., bleu 1.000
he's calm . => il est paresseux ., bleu 0.658
```

Run 4: `loss 0.018, 6071.4 tokens/sec on cpu`
```
go . => vas-y !, bleu 0.000
i lost . => je <unk> ., bleu 0.000
i'm home . => je suis partie ., bleu 0.512
he's calm . => <unk> ., bleu 0.000
```

Run 5: `loss 0.017, 4642.7 tokens/sec on cpu`
```
go . => va !, bleu 1.000
i lost . => j'ai perdu ., bleu 1.000
i'm home . => je suis chez <unk> ., bleu 0.752
he's calm . => je suis calme !, bleu 0.000
```

I guess there is indeed some issue with PT tranformer chapter since seq2seq has comparable results to mx but the transformer results are not very consistent. Thanks for flagging this @astonzhang , I'll try to find and fix the issue. @ypandya Would be nice if you can also take a look. Thanks!

**Comment by astonzhang** [comment]
Just one more example of inconsistent performance

![Screen Shot 2020-10-22 at 12 59 49 PM](https://user-images.githubusercontent.com/22279212/96923708-85445600-1466-11eb-8751-1c456b3bab99.png)
![Screen Shot 2020-10-22 at 12 59 43 PM](https://user-images.githubusercontent.com/22279212/96923714-86758300-1466-11eb-8e8c-61b4a91b6d89.png)

**Comment by 315930399** [comment]
First, *torch.repeat* should be used instead of *torch.repeat_interleave*. Someone point this out in the discussion of https://d2l.ai/chapter_attention-mechanisms/transformer.html
Second, I believe there are some issues in the *d2l.torch.predict_s2s_ch9* since it outputs differently with the same input. The mxnet version don't have this problem.

**Comment by AnirudhDagar** [comment]
Please see my comments on #1528 for the bug fix and reasoning.
Once the PR is merged, it will trigger a close on this issue. Thanks!

**Comment by astonzhang** [comment]
Now both look better to me!

![Screen Shot 2020-11-20 at 4 28 00 PM](https://user-images.githubusercontent.com/22279212/99862097-61e7f600-2b4d-11eb-9621-d8f6a7ee0494.png)

![Screen Shot 2020-11-20 at 4 28 03 PM](https://user-images.githubusercontent.com/22279212/99862102-64e2e680-2b4d-11eb-9781-210cd89d3b43.png)
